runtime/memprof.c

```diff
@@ -16,23 +16,531 @@
+ *   possible to reconstruct the allocation addresses of the tracking
+ *   entries, so they are simply deleted. The new profile effectively
+ *   begins with the following allocation.
+ *
+ * Most of this complexity is managed in caml_memprof_track_young().
+ *
+ * 7. Callbacks
+ *
+ * Some callbacks are run at allocation time, for allocations from
+ * Caml (see under "Sampling" above). Other allocation callbacks, and
+ * all post-allocation callbacks, are run during
+ * `caml_memprof_run_callbacks_exn()`, which is called by the
+ * runtime's general pending-action mechanism at poll points.
+ *
+ * We set the domain's action-pending flag when we notice we have
+ * pending callbacks. Caml drops into the runtime at a poll point, and
+ * calls `caml_memprof_run_callbacks_exn()`, whenever the
+ * action-pending flag is set, whether or not memprof set it. So
+ * memprof maintains its own per-domain `pending` flag, to avoid
+ * suspending/unsuspending sampling, and checking all the entries
+ * tables, when there are no pending callbacks.
+ *
+ * This is particularly important because when we unsuspend sampling,
+ * we reset the young-limit, which has the side-effect of setting the
+ * domain's action-pending flag. TODO: consider changing
+ * `caml_reset_young_limit` so it doesn't do this.
+ *
+ * Allocation callbacks are always run by the thread which made the
+ * allocation, unless that thread terminates before running the
+ * callback, in which case it is inherited by the domain.
+ *
+ * Callbacks are run by iterating through candidate entries in a entry
+ * table. See under "Entries" above. A single entry may have more than
+ * one callback to run (if, for example, it has been promoted *and*
+ * garbage collected since the last time callbacks for that entry were
+ * run) - they are run in the natural order.
+ */
+
+/* number of random variables in a batch */
+#define RAND_BLOCK_SIZE 64
 
 /* type aliases for the hierarchy of structures for managing memprof status. */
 
+typedef struct entry_s entry_s, *entry_t;
+typedef struct entries_s entries_s, *entries_t;
 typedef struct memprof_domain_s memprof_domain_s, *memprof_domain_t;
 typedef struct memprof_thread_s memprof_thread_s, *memprof_thread_t;
+typedef struct memprof_orphan_table_s memprof_orphan_table_s,
+  *memprof_orphan_table_t;
+
+/* [Gc.Memprof.allocation_source] */
+
+/* At present (since OCaml 5), SRC_MARSHAL can't be produced, because
+ * unmarshalling uses the regular allocation functions. */
+
+enum { SRC_NORMAL = 0, SRC_MARSHAL = 1, SRC_CUSTOM = 2 };
+
+/* A memprof configuration is held in an object on the Caml heap, of
+ * type Gc.Memprof.t. Here we define getter macros for each field, and
+ * a setter macro for the status field (which is updated). */
+
+#define CONFIG_FIELDS 9
+
+#define CONFIG_FIELD_STATUS        0
+#define CONFIG_FIELD_LAMBDA        1
+#define CONFIG_FIELD_1LOG1ML       2
+#define CONFIG_FIELD_FRAMES        3
```

Comment:
"FRAMES" is a bit abstract to me.
```suggestion
#define CONFIG_FIELD_STACK_FRAMES  3
```